lib/filter-schema.js

```diff
@@ -0,0 +1,163 @@
+
+    // the filter hook must be the last one to be executed (after all the authContextHook ones)
+    app.ready(err => {
+      /* istanbul ignore next */
+      if (err) throw err
+      app.graphql.addHook('preExecution', filterGraphQLSchemaHook(namespace).bind(app))
+    })
+  }
+
+  if (app[kDirectiveNamespace][namespace]) {
+    app[kDirectiveNamespace][namespace].push({
+      policy,
+      policyFunction
+    })
+  } else {
+    app[kDirectiveNamespace][namespace] = [{
+      policy,
+      policyFunction
+    }]
+  }
+}
+
+function filterGraphQLSchemaHook (namespace) {
+  return async function filterHook (schema, document, context) {
+    if (!isIntrospection(document)) {
+      return
+    }
+
+    const filteredSchema = await filterSchema(schema,
+      this[kDirectiveNamespace][namespace],
+      context)
+
+    return {
+      schema: filteredSchema
+    }
+  }
+}
+
+function isIntrospection (document) {
+  for (const queryType of document.definitions) {
+    if (queryType.operation !== 'query') {
+      // if there is a mutation or subscription, we can skip the introspection check
+      break
+    }
+
+    // if there is an introspection operation, we must filter the schema
+    if (queryType.selectionSet.selections.some(sel => (
+      sel.name.value === '__schema' ||
+      sel.name.value === '__type'
+    ))) {
+      return true
+    }
+  }
+  return false
+}
+
+async function filterSchema (graphQLSchema, policies, context) {
+  const filterDirectiveMap = {}
+
+  // each `policies` item is a directive
+  let skipFiltering = true
+  for (const { policy, policyFunction } of policies) {
+    // each `policy` contains all the GraphQL OBJECT and FIELDS that are affected by the directive
+    for (const [typeName, typePolicy] of Object.entries(policy)) {
+      // different `policies` item can affect the same GraphQL OBJECT
+      if (filterDirectiveMap[typeName] === undefined) {
+        filterDirectiveMap[typeName] = {}
+      } else if (filterDirectiveMap[typeName] === false) {
+        // if the object has already been filtered, we can skip the field processing
+        continue
+      }
+
+      const schemaType = graphQLSchema.getType(typeName)
+      const schemaTypeFields = typeof schemaType.getFields === 'function'
+        ? schemaType.getFields()
+        : {}
+      for (const [fieldName, fieldPolicy] of Object.entries(typePolicy)) {
+        // each `fieldName` is a single GraphQL item associated with the directive
+
+        if (filterDirectiveMap[typeName] === false || filterDirectiveMap[typeName][fieldName] === false) {
+          // if we have already decided to filter out this field
+          // it does not need to be checked again
+          continue
+        }
+
+        let canShowDirectiveField = true
+        const isObjectPolicy = fieldName === '__typePolicy'
+        try {
+          // https://github.com/graphql/graphql-js/blob/main/src/type/definition.ts#L974
+          const info = {
```

Comment:
I'm trying to create the `GraphQLResolveInfo` that contains information about the operation's execution state, including the field name.

I did not cover all the cases, but it will be slightly different from the `info` argument received as input by the resolver function.

Any suggestion to recreate it is highly appreciated

Reply:
Apologies for the slow response, I've been very busy this week - I will have a look this weekend if that's okay!

Reply:
There are a couple of functions exported from the `graphql` package that might be of interest here:
- `buildExecutionContext`
- `buildResolveInfo` (requires an execution context)

Reply:
I did not find a way to build the info object actually.

Would you mind defining a subset info object in the first place?

Reply:
What would this subset include? Do you anticipate this being problematic for the type definition of the `GraphQLResolveInfo` type?

Depending on the data included, I think this is okay for the initial release of this feature if it's not possible - we'd need to:
- Make this clear in the documentation

What was blocking you on this out of interest?

Reply:
Let me make an example:
if you assign the `@auth` directive to a `type MyObject` and you will set this type as output for two different `Query`, the `info` object depends on the query that is being executed - the resolver changes.

Instead, during the filtering, the `applyPolicy` function is not being executed as a resolver wrapper, but it is executed on the `MyObject` type without any other context.

For this reason I was unable to build the same argument